### ISSUE TITLE
Navigation menu block: Explore horizontal move controls

### ISSUE THREAD

**Issue by sarahmonster** [WordPress/gutenberg#17093]
From #16821 

> Refine the horizontally aligned inner blocks.
> Explore how this pattern can extend to other blocks like button, columns, etc.
> Improve how child blocks perform in narrow spaces.

## Exploring the problem

**Problem:** 

Since the block interface has primarily been designed with a stacked vertical alignment in mind, there isn't a good mechanism for manipulating items that are aligned side-by-side. Users need a way of manipulating blocks when they're horizontally, rather than vertically, aligned.

**How we determine if the solution has succeeded:** 

@mtias @mapk @karmatosed How did you want to select the best solution here? Some ideas:
- Run lightweight usability testing
- Take a group consensus / dot voting
- Determine a decider responsible for selecting the best path forward

**Potential roadblocks:**

We'll want to design this first and foremost for users who don't have access to drag and drop. How do we allow keyboard and screen reader users to easily rearrange elements without drag-and-drop? Drag and drop controls should still be available as an additive functionality.

Controls need to be large enough to allow for reasonable touch target size. What's "reasonable" differs in context, so let's aim to at least provide what the vertical movers allow for, which looks to be 28×24 on desktop, 36x30 on mobile. 

Obviously we'll also need to apply this pattern to mobile interfaces as well, which adds an additional layer of complexity.

We'll also want to consider these controls insofar as they impact #16580, since this may determine how we move forward here.

## Original controls

Our original explorations centred primarily around the placement of the arrow mover controls. Due to time constraints, these explorations were shared only in Figma and in Slack discussions, so I'm revisiting those explorations now. 

For reference, this is the pattern that we eventually landed on:

<img width="324" alt="Screenshot 2019-08-19 at 10 24 20" src="https://user-images.githubusercontent.com/376315/63273203-87f85980-c26b-11e9-9ddb-1a9eb8c14ddb.png">

We also included a whole set of advanced move controls, to take into account accessibility needs and allow for quicker reordering of potentially complex menus:

<img width="311" alt="Screenshot 2019-08-19 at 16 47 35" src="https://user-images.githubusercontent.com/376315/63298272-1935f300-c2a1-11e9-9e91-ae850f9e07e8.png">

Usability testing indicated that this was a pretty straightforward pattern. Both new and existing Gutenberg users were able to easily understand how to move and rearrange their menu items.

## Explorations

I've updated our original explorations using the new menu block design from #16974, and combining with some of the explorations from #16580. 

### Current (no background) styling:
![navigation-menu-mover-placement](https://user-images.githubusercontent.com/376315/63297451-6ca74180-c29f-11e9-8401-0a498cfd2fd3.png)

### Border-box styling:
![navigation-menu-mover-placement-border](https://user-images.githubusercontent.com/376315/63297449-6c0eab00-c29f-11e9-9832-66136c3d1662.png)

### Inversed:
![navigation-menu-mover-placement-inverse](https://user-images.githubusercontent.com/376315/63298172-e7249100-c2a0-11e9-9a78-d972f1b18e67.png)

### Mobile is pretty straightforward:
![vertical-block-mover-mobile](https://user-images.githubusercontent.com/376315/63297454-6ca74180-c29f-11e9-917f-3efc80a99880.png)

## Your feedback

1. Which of these options seems like it would be most intuitive for users?
2. Are there alternative options we should consider here?
3. Are there accessibility concerns with any of these solutions?
4. Does consideration of horizontal movers impact our approach to #16580 at all?
5. Do we want to provide a mechanism for more complex block movement options, as per the original designs?

## Next steps

Once we've narrowed the explorations down a bit, we can investigate how those different options might work when applied to different blocks to better test these approaches. The styling of the mover buttons is likely to impact which layout options work best here though, so we may want to wait until #16580 has a clear resolution before moving forward.

From #16821, we'll also want to look at
> Improve how child blocks perform in narrow spaces.

@mtias, @mapk could you clarify what you mean here so I can be sure to address it?

**Comment by ZebulanStanphill** [comment]
Splitting the mover controls seems like a bad idea from an accessibility tab-order viewpoint in my opinion, and in most cases it just looks weird.

Having the controls stick out like a flag from the left border of the block also feels weird.

I think it would make the most sense to either use the same position as vertical movers (similar to how the position would be the same for either type on mobile), or put the controls on the bottom.

**Comment by karmatosed** [comment]
I have a few thoughts buzzing around about this but going to start by answering what you outline for feedback:

> Which of these options seems like it would be most intuitive for users?

I have to say all of them for me add a cognitive load to the experience, I don't want to choose one right now as think we can explore through why we are doing this and how. More in next answer.

> Are there alternative options we should consider here?

I wonder if we need to explore how drag and drop could work or layering controls. As you are moving do you also need to see the other controls? I wonder if you don't. Here is a super rough sketch, but what if as you were in moving state, the toolbar changed?

![image](https://user-images.githubusercontent.com/253067/63624296-e028b600-c5f3-11e9-8ffd-caaae81147e7.png)

> Do we want to provide a mechanism for more complex block movement options, as per the original designs?

I think we should consider how simple this could be, or at least what can we boil it down to and iterate?

**Comment by sarahmonster** [comment]
Following a [discussion in Slack today](https://wordpress.slack.com/archives/C02S78ZAL/p1567084252106700) (Note: You may need a [Slack account](https://make.wordpress.org/chat/) to log in.), it seems that none of the options presented are good to work from. 

I was working from the assumption that we'd want to reuse the existing control pattern used by horizontal movers, but we clarified that that isn't the case. Instead, we want to try introducing a _completely new_ pattern for horizontal move controls.

Given that time is of the essence here, we agreed to work from these mockups:

![Buttons with arrows over a navigation menu item](https://user-images.githubusercontent.com/376315/63948568-e6c98a00-ca46-11e9-8e8b-a06f9e65bfff.png)

![Buttons with arrows over an image](https://user-images.githubusercontent.com/376315/63948567-e6c98a00-ca46-11e9-85ee-8ad91e94b8ef.png)


and iterate directly in code. (cc @draganescu)

If that doesn't work and we need to go back to the drawing board. Here are some suggestions for areas to explore further, if needs be:
- Put controls into the ellipsis menu.
- Put controls in the toolbar.
- Put controls inside of the block itself (inline).

For the moment, since we're going to move directly to prototyping and iterating in code, this issue can be closed.
